accounts/abi/bind/bind.go

```diff
@@ -219,18 +228,66 @@ func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]
 			}
 		}
 	}
+
 	// Check if that type has already been identified as a library
 	for i := 0; i < len(types); i++ {
 		_, ok := isLib[types[i]]
 		contracts[types[i]].Library = ok
 	}
-	// Generate the contract template data content and render it
-	data := &tmplData{
-		Package:   pkg,
-		Contracts: contracts,
-		Libraries: libs,
-		Structs:   structs,
+
+	var data interface{}
+
+	if isPrecompile {
+		templateSource = tmplSourcePrecompileGo
+		if lang == LangJava {
+			return "", errors.New("java binding for precompiled contracts is not supported yet")
+		}
+
+		if len(contracts) != 1 {
+			return "", errors.New("cannot generate more than 1 contract")
+		}
+		precompileType := types[0]
+		firstContract := contracts[precompileType]
+
+		funcs := make(map[string]*tmplMethod)
+
+		for k, v := range firstContract.Transacts {
+			funcs[k] = v
+		}
+
+		for k, v := range firstContract.Calls {
+			funcs[k] = v
+		}
+
+		_, isAllowList := funcs[readAllowListFuncKey]
+		if isAllowList {
+			// remove these functions as we will directly inherit AllowList
+			delete(funcs, readAllowListFuncKey)
+			delete(funcs, setAdminFuncKey)
+			delete(funcs, setEnabledFuncKey)
+			delete(funcs, setNoneFuncKey)
+		}
+
+		precompileContract := &tmplPrecompileContract{
+			tmplContract: firstContract,
+			AllowList:    isAllowList,
+			Funcs:        funcs,
+		}
+
+		data = &tmplPrecompileData{
+			Contract: precompileContract,
+			Structs:  structs,
+		}
```

Comment:
nit: may be easier to follow if we made this part into a function and got `data` as the return value (or `data` and `templateSource`)